[openstack] Config drive option

zuzzas opened the pull request:
**What this PR does / why we need it**:

Add an option to use config drive while provisioning OpenStack Servers.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator|developer
-->
```improvement operator
Added an option to use configDrive in the OpenStackMachineClass
```

hardikdr commented:
Thanks for the PR @zuzzas.

hardikdr commented:
@afritzler @kayrus @dguendisch Would be really nice if you could take a brief look if possible :)

kayrus commented on pkg/driver/driver_openstack.go:
```diff
@@ -160,6 +161,29 @@ func (d *OpenStackDriver) Create() (string, string, error) {
 
 	metrics.APIRequestCount.With(prometheus.Labels{"provider": "openstack", "service": "nova"}).Inc()
 
+	var useConfigDrive *bool
```
I'd move the logic below into a separate func. However the whole `Create` func already consists of `if then else`. Apart from the above, it's better to define `var useConfigDrive bool` here and use `ConfigDrive:      &useConfigDrive,` in `createOpts definition.

kayrus commented on pkg/driver/driver_openstack.go:
```diff
@@ -160,6 +161,29 @@ func (d *OpenStackDriver) Create() (string, string, error) {
+		for _, subnet := range allSubnetsForMainNet {
+			if !subnet.EnableDHCP {
+				trueVal := true
+				useConfigDrive = &trueVal
```
```suggestion
				useConfigDrive = true
				break
				
```

zuzzas replied:
Removed auto-detection.

kayrus commented on pkg/driver/driver_openstack.go:
```diff
@@ -160,6 +161,29 @@ func (d *OpenStackDriver) Create() (string, string, error) {
 	metrics.APIRequestCount.With(prometheus.Labels{"provider": "openstack", "service": "nova"}).Inc()
 
+	var useConfigDrive *bool
+	if len(serverNetworks) > 0 {
```
Does it make sent to loop over all serverNetworks?
```suggestion
	for _, serverNetwork := range serverNetworks {
```

zuzzas replied:
I've removed all the auto-detection logic due to this. We can't make such assumptions.

kayrus commented on pkg/driver/driver_openstack.go:
```diff
@@ -606,3 +631,13 @@ func resourceInstanceBlockDevicesV2(rootDiskSize int, imageID string) ([]bootfro
 	return blockDeviceOpts, nil
 }
+
+func dhcpDisabled(subnets []subnets.Subnet) bool {
```
I don't see that this func is used.

zuzzas replied:
Done.

afritzler commented:
One other reason for explicit support in the `OpenStackMachineClass`: we had in the past the situation where the `dhcp` interfaces where present in the subnet. However they became unresponsive. We already thought back where it would be good to have the `ConfigDrive` support as an alternative available, but didn't follow through.

zuzzas commented:
@afritzler 
I've removed all the auto-detection logic and switched to a simple config option. It was nice, but it relied on some assumptions (as pointed by https://github.com/gardener/machine-controller-manager/pull/457#discussion_r418215596) that can clearly break user's expectations.

zuzzas commented on pkg/driver/driver_openstack.go:
```diff
@@ -25,10 +25,9 @@ import (
-
 	"github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
 	"github.com/gardener/machine-controller-manager/pkg/metrics"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
```
This was done by `goimports`.

hardikdr commented:
@afritzler It Would be nice if you could take a look again.

hardikdr commented:
There seems to be a minor file-conflict, can you please take a look @zuzzas .
If there are no more objections, I'll take a final look and prefer to merge it soon.

zuzzas commented:
@hardikdr, it is done.

hardikdr commented:
@kayrus @afritzler I would be nice if you could take a look and approve if possible :)

afritzler commented:
Let me test that against our environment and get back to you by Monday. Pretty swamped atm.

prashanth26 commented:
Hi @afritzler ,

Did you get a chance to test these changes?

hardikdr commented:
cc @dkistner @MartinWeindel

MartinWeindel commented:
@zuzzas Can use adjust the PR description to reflect the latest state? There is still described that automatic detection is used, but it's not in the code anymore.

afritzler commented:
Ok, I finally also came around to test this PR. Creating machines without and with config drive support
```
useConfigDrive: true
```
in the `MachineClass` is working on our OpenStack environment.

/lgtm

zuzzas commented:
>@zuzzas Can use adjust the PR description to reflect the latest state? There is still described that automatic detection is used, but it's not in the code anymore.

Done.

hardikdr commented:
Thanks, everyone. I will go ahead merging then.
